### DISCUSSION TITLE
How to preserve the original error information whem mapping them?

### DISCUSSION THREAD

**Discussion by diogob** [seasonedcc/composable-functions#157] [Ideas]
We usually transform errors in composables using `mapError`.
I see a very common pattern:

```typescript
const composableWithMorePreciseError = mapErrors(
  composableWithGenericError, 
  (errors) => [...errors, new MySpecificError()]
)
```

The goal above is not so much to transform an error, but to add some additional context or pre-processing before sending the errors elsewhere.

## The problem

It is very common for developers to skip the spread, and remove existing errors from the list when doing the above. So the code above would be naively written as:

```typescript
const composableWithMorePreciseError = mapErrors(
  composableWithGenericError, 
  (errors) => [new MySpecificError()]
)
```

Which is also an implementation I have seen in the wild. And hiding these details, despite looking like a good idea, hinder troubleshooting after the fact as it is  very hard to know what we will need to debug a live system.

Perhaps we should create an incentive, in our API, for preserving the original errors.

## A possible solution

A new combinator such as `appendErrors : (errors: Error[]) => Error[]`.

This combinator, has the same signature as `mapError` but it just appends the errors returned to the original list. So we could write the simpler version of the examples above and still keep the full error history:

```typescript
const composableWithMorePreciseError = appendErrors(
  composableWithGenericError, 
  (errors) => [new MySpecificError()]
)
```

**Comment by diogob** [comment]
@gustavoguichard @danielweinmann opinions?

**Reply by gustavoguichard** [reply]
I'm not against it, but it is not a use case I need to cover.
It is also one of those combinators we have enough primitives to build in our own projects, but - despite being easy - it adds one more function to maintain if added to the repo.
Your call

**Reply by diogob** [reply]
I agree that it is simple enough to delegate to users (and perhaps have in some recipes section of the docs). However, I still wonder if nudging the user towards good patterns wouldn't warrant an out of the box combinator.

**Reply by diogob** [reply]
You see, I'm not really concerned with the developer that masters composables and is willing to write a custom combinator. This idea is geared towards the (potentially newbie dev) newcomer to the library.

**Reply by danielweinmann** [reply]
I think a newcomer wouldn't see the future benefit of using `appendErrors` and would probably choose `mapErrors` because the name is closer to what they are looking for. I never came across this use case also. Maybe let's try documenting `mapErrors` better before adding a new combinator?

**Reply by danielweinmann** [reply]
BTW @diogob, I see you used `mapError` in your example. I'm only familiar with `mapErrors`. Was it a typo?

**Reply by diogob** [reply]
typo fixed

**Comment by diogob** [comment]
I'll close this, documentation seems to be the way to go
